### PR TITLE
Initial Interfaces

### PR DESCRIPTION
**Pull request by XedinUnknown**

### CODE REVIEW

**Comment by XedinUnknown** [comment]
What the `CodeInvokerInterface#invoke()` really reminds me of is `EventManagerInterface#trigger()` - except the fact, which is the problem, that `trigger()` accepts the args as the 3rd parameter, while `CodeInvokerInterface` accepts the args as the 2nd parameter.

**Comment by XedinUnknown** [comment]
Another idea is to make the first argument `mixed` in a new super-interface `InvokerInterface`, and re-name it to `$subject`

*Edit*: This can be based on the idea that conceptually anything could be invoked, and how to do that depends on the implementation. This can also be further narrowed down by more specialized interfaces; in fact, this will probably be the most common case. However, it can still be useful in dealing with exceptions, and in more generalized algorithms.

File: src/CodeInvokerInterface.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+namespace Dhii\Invocation;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+
+/**
+ * Something that can invoke commands by code.
+ *
+ * @since [*next-version*]
+ */
+interface CodeInvokerInterface
```
Comment (mecha) [inline src/CodeInvokerInterface.php]: The name `CodeInvokerInterface` is a bit misleading, mainly due to the use of the term "code" which is ambiguous (program code vs key-like code). Consider using the term "command".

**Comment by mecha** [comment]
> `trigger()` accepts the args as the 3rd parameter, while `CodeInvokerInterface` accepts the args as the 3rd parameter

I think you meant `CodeInvokerInterface` accepts the args as the *2nd parameter*.

----

Re. `InvokerInterface` and `$subject`:

The only thing that worries me is the parameter type for `$subject`. I'd like to not use `mixed` as much as possible. Reason: when a consumer depends on an `InvokerInterface` to invoke something, from the consumer's POV the invocation has undefined behavior, because the invoker accepts _anything_ but promises _nothing_ with regards to how the input is treated and how it leads to invocation - which is not desirable.

What _does_ make sense is to depend on an invoker interface for something specific: `CodeInvokerInterface` (accept code, map it to some invocable thing), `CallableInvokerInterface` (accept callable, invoke it), `PhpInvokerInterface` (accept PHP code, eval it), etc...

My conclusion is that a super-interface does not make much sense.

**Comment by XedinUnknown** [comment]
Yes, my mistake WRT parameter.

---

All of them invokers, but have no common ancestor. *That* doesn't make much sense to me. You could have an invoker that knows a module loader to have a `load()` method, and a `callable` object to have an `__invoke()`. All of them could then be "invoked" using the invoker, in a uniform way. This is cool and flexible. Another useful consequence is that you can do mapping between named and ordered parameters this way.

**Comment by mecha** [comment]
Not having a common ancestor is perfectly normal.

By having a common ancestor, you are implying that _all_ invokers, both those that implement the ancestor interface and those that implement the specific interface, must be able to invoke _anything_. This makes the specific interfaces useless, and the ancestor interface more so.

If that was unclear:

```php
interface InvokerInterface
{
    /** @param mixed $subject */
    public function invoke($subject, array $args =[]);
}

interface CodeInvokerInterface extends InvokerInterface
{
    /** @param string|Stringable $code */
    public function invoke($code, array $args = []);
}

class MyInvoker implements CodeInvokerInterface
{
    // ...
}

class MyConsumingClass
{
    public function foobar(InvokerInterface $invoker)
    {
        $invoker->invoke($this->getSubject());
    }
}

$consumer = new MyConsumingClass();
$invoker     = new MyInvoker();

$consumer->foobar($invoker);
```

The invoker class must be able to invoke any subject, just like its ancestor. Substitution with a class that implements a more derived class results in undefined behavior. What happens now? Any deviation from the intended  behavior of an implementation of the super interface breaks substition.

**Comment by XedinUnknown** [comment]
The invoker that accepts `mixed` is not required to be able to invoke anything. It's required to be able to accept and work with anything. But it may fail, because not anything *can* be invoked. And that's alright.

**Comment by mecha** [comment]
IMO, it's not alright.

An invoker interface is meant to provide an abstraction for all invokers. Consumers can use this abstraction to not bother about the implementation. At the same time, the consumer is already aware of what it wants to achieve. It makes no sense to swap out a `CodeInvokerInterface` for a `CallableInvokerInterface` if the consumer is passing the invoker a code and not a callable. From the consumers POV, the invocation should fail if the code did not correspond to an invocable thing, or if the invocation itself encountered a problem. Not because the input is invalid. For this reason, I think that depending on a generic `InvokerInterface` is an indication of a design flaw.

But, an object failing because the input is invalid is OK - it is an indication of misuse of the object's API or an unhandled case in the application logic. Using `mixed` however raises the question in the consumer's mind: "What _IS_ invalid input"? To which there is no answer. This is what I mean by undefined behavior. Conversely, it's OK if the consumer knows that it should pass, say, a `string` or `Stringable`. The types are clearly defined and by applying the same idea to the dependency recursively, behavior is also well defined. This is *strong behavioral sub-typing* which LSP is based on.

File: src/Exception/InvocationFailureExceptionInterface.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Dhii\Invocation\Exception;
+
+/**
+ * An exception that occurs when an invoker fails to invoke a command.
+ *
+ * @since [*next-version*]
+ */
+interface InvocationFailureExceptionInterface extends CommandInvokerExceptionInterface
```
Comment (mecha) [inline src/Exception/InvocationFailureExceptionInterface.php]: Shouldn't this exception also be aware of the command code, and perhaps even the arguments?

File: src/CommandInvokerInterface.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /**
+     * Invokes a command.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable $code   The command to invoke.
+     * @param array             $params The parameters to pass to the command.
+     */
+    public function invoke($code, array $params = []);
```
Comment (mecha) [inline src/CommandInvokerInterface.php]: The second parameter accepts arguments, not parameters. Use `$args` instead.

**Review by mecha — CHANGES_REQUESTED** [review]

**Review by mecha — APPROVED** [review]
LGTM after the latest changes

**Pull request merged**